lib/blinkist-airbrake-scrubber/hash.rb

```diff
@@ -0,0 +1,22 @@
+# Add Hash#traverse(&block)
+class Hash
+
+  # Traverse hash and run block on each key-value pair
+  # Block should return [key, value] array
+  def traverse(&block)
+    inject({}) do |memo, (key,value)|
+      # Go deep
+      if Hash === value
+        value = value.traverse(&block)
+      elsif value.respond_to?(:to_hash)
+        value = value.to_hash.traverse(&block)
+      end
+
+      # Process key-value pair
+      new_key, new_value = block.call key, value
+      memo[new_key]      = new_value
+
+      memo
+    end
```

Comment:
This method is nice, but also insufficient for our purposes, since it fails to catch for example:

```ruby
{ params: [ { email: "a@a.com" }, { email: "b@b.com" } ] }
```

Reply:
An alternative (probably a bit buggy):
```ruby
class DeepTraversal
  def initialize(source)
    @source = source
  end

  def traverse(&block)
    recursive_traverse(@source, &block)
  end

  private

  def recursive_traverse(input, &block)
    case input
    when Array
      input.map { |i| recursive_traverse(i, &block) }
    when Hash
      Hash[input.map { |key, value|
        case value
        when String
          [key, block.call(key, value)]
        else
          [key, recursive_traverse(value, &block)]
        end
      }]
    else
      input
    end
  end
end

source = { params: [ { email: "a@a.com" }, { email: "b@b.com" }, {x: "blah"} ], email: "c@c.com", a: 1 }

target = DeepTraversal.new(source).traverse do |key, value|
  if key == :email
    "[Filtered]"
  else
    value
  end
end

pp source
pp target
```

Reply:
Okay, but do we anywhere in Airbrake have such case?

Reply:
I would bet about 3.50€ that we have something like this hiding somewhere

Reply:
That's a helluva lot of money, mister! :P But I like this idea, since it would definitely make the scrubbing more sturdy, gonna try to get that :-)